sphinx/directives/code.py

```diff
@@ -156,6 +154,190 @@ def run(self):
         return [literal]
 
 
+class LiteralIncludeReader(object):
+    INVALID_OPTIONS_PAIR = [
+        ('pyobject', 'lines'),
+        ('lineno-match', 'lineno-start'),
+        ('lineno-match', 'append'),
+        ('lineno-match', 'prepend'),
+        ('start-after', 'start-at'),
+        ('end-before', 'end-at'),
+        ('diff', 'pyobject'),
+        ('diff', 'lineno-start'),
+        ('diff', 'lineno-match'),
+        ('diff', 'lines'),
+        ('diff', 'start-after'),
+        ('diff', 'end-before'),
+        ('diff', 'start-at'),
+        ('diff', 'end-at'),
+    ]
+
+    def __init__(self, filename, options, config):
+        # type: (unicode, Dict, Config) -> None
+        self.filename = filename
+        self.options = options
+        self.encoding = options.get('encoding', config.source_encoding)
+        self.lineno_start = self.options.get('lineno-start', 1)
+
+        self.parse_options()
+
+    def parse_options(self):
+        # type: () -> None
+        for option1, option2 in self.INVALID_OPTIONS_PAIR:
+            if option1 in self.options and option2 in self.options:
+                raise ValueError(_('Cannot use both "%s" and "%s" options') %
+                                 (option1, option2))
+
+    def read_file(self, filename):
+        # type: (unicode) -> List[unicode]
+        try:
+            with codecs.open(filename, 'r', self.encoding, errors='strict') as f:  # type: ignore  # NOQA
+                text = f.read()  # type: unicode
+                if 'tab-width' in self.options:
+                    text = text.expandtabs(self.options['tab-width'])
```

Comment:
I modified the behavior of `:tab-width:` option. Because it was invoked at final stage of `literalinclude` directive.
But it conflicts with `:dedent:` option. The `dedent` option skips a TAB as a 1 character (not expanded). I feel this is a bit strange.